Updates for 12. Data Tables (#465)
atbcb/ICTTestingBaseline: _baselines/12DataTables.md

-   In the Baseline Test instructions, where an ARIA role is identified, it is the first valid ARIA role attribute value.

### 12.A Test Procedure for Data Table Roles
**Baseline Test ID:** 12.A-DataTableRole

#### Identify Content
<p id="12aIC">Content/data that is visually presented in a table, arranged in rows and columns, where the content is not in a meaningful sequence when linearized.</p>

<p>Note: Linearization of table content is the presentation of a table’s two-dimensional content in one-dimensional order of the content in the source, beginning with the first cell in the first row and ending with the last cell in the last row, from left to right, top to bottom.</p>

#### Test Instructions
<ol id="12aTI">

>     <li id="12aTI-1">Table: Check the data table has programmatic role of table using one of the following techniques. If more than one technique is used, select one and perform the remaining tests using that selection. [SC 4.1.2]:

Reviewer: ```suggestion
    <li id="12aTI-1">Table: Check that the data table has a programmatic role of "table" using one of the following techniques. If more than one technique is used, select the first role explicitly defined and perform the remaining tests using that selection. [SC 4.1.2]:
```

        <ul>

>             <li>HTML <code>&lt;table&gt;</code> that does not have a role attribute that changes its role from table, such as <code>role="presentation"</code> or <code>role="none"</code>.</li>

Reviewer: ```suggestion
            <li>HTML <code>&lt;table&gt;</code> that does not have an explicitly defined role attribute that changes its role from table, such as <code>role="presentation"</code> or <code>role="none"</code>.</li>
```

            <li>ARIA <code>role="table"</code></li>
            <li>ARIA <code>role="grid"</code></li>
            <li>ARIA <code>role="treegrid"</code></li>
        </ul>
    </li>

>     <li id="12aTI-2">Table data cell: For the technique used, check that each data cell is programmatically assigned a role of cell and within a row [SC 4.1.2]:

Reviewer: ```suggestion
    <li id="12aTI-2">Table data cell: For the technique used, check that each data cell is programmatically assigned a role of cell and within a parent element with a role of row [SC 4.1.2]:
```

        <ul>

>             <li>For HTML <code>&lt;table&gt;</code>: the <code>&lt;td&gt;</code> cell must be within a <code>&lt;tr&gt;</code> row.</li>

Reviewer: ```suggestion
            <li>For HTML <code>&lt;table&gt;</code>: the <code>&lt;td&gt;</code> cell must be within a parent <code>&lt;tr&gt;</code> row element.</li>
```

            <li>For ARIA <code>role="table"</code>: the data cell element with <code>role="cell"</code> must be within an element with <code>role="row"</code>.</li>

>             <li>For ARIA <code>role="grid"</code> or ARIA <code>role="treegrid"</code>: the data cell element with <code>role="gridcell"</code> is within an element with <code>role="row"</code>.</li>

Reviewer: ```suggestion
            <li>For ARIA <code>role="grid"</code> or ARIA <code>role="treegrid"</code>: the data cell element with <code>role="gridcell"</code> is within a parent element with <code>role="row"</code>.</li>
```

        </ul>
    </li>

>     <li id="12aTI-3">Table header cell: For the technique used, check that each header cell is programmatically assigned a role of header and within a row [SC 4.1.2]:

Reviewer: ```suggestion
    <li id="12aTI-3">Table header cell: For the technique used, check that each header cell is programmatically assigned a role of header and within a parent row element [SC 4.1.2]:
```

        <ul>

>             <li>For HTML <code>&lt;table&gt;</code>: the <code>&lt;th&gt;</code> header must be within a <code>&lt;tr&gt;</code> row.</li>

Reviewer: ```suggestion
            <li>For HTML <code>&lt;table&gt;</code>: the <code>&lt;th&gt;</code> header must be within a parent <code>&lt;tr&gt;</code> row element.</li>
```

>             <li>For an ARIA <code>role="table"</code>, ARIA <code>role="grid"</code> or <code>role="gridtree"</code>: each element with <code>role="columnheader"</code> or <code>role="rowheader"</code> must be within an element with <code>role="row"</code>.</li>

Reviewer: ```suggestion
            <li>For an ARIA <code>role="table"</code>, ARIA <code>role="grid"</code> or <code>role="gridtree"</code>: each element with <code>role="columnheader"</code> or <code>role="rowheader"</code> must be within a parent element with <code>role="row"</code>.</li>
```

        </ul>
    </li>
</ol>

#### Test Results
<p id="12aTR">If any of the above tests fail, Baseline Test 12.A-DataTableRole fails.</p>

### 12.B Test Procedure for Data Table Header Association
**Baseline Test ID:** 12.B-DataTableHeaderAssociation

#### Identify Content
<p id="12bIC">For any data table identified in 12.A, identify all column and row headers for each data cell.</p>

<ol>
    <li id="12bTI-1">Data cell to header(s) association: Use the programmatic technique (HTML or ARIA) identified in 12.A. Check that data cells are programmatically associated with its headers [SC 1.3.1]:
        <ul>
            <li id="12bTI-1a">For a simple HTML <code>&lt;table&gt;</code> only: with column headers only in the first row that apply to all data cells in the same column and row headers only in the first column that apply to all data cells in the same row, each header cell may be marked up with <code>&lt;th&gt;</code> without additional attributes.</li>

>             <li id="12bTI-1b">For any HTML <code>&lt;table&gt;</code>: a header cell may be marked up with <code>&lt;th scope="col|row|colgroup|rowgroup"&gt;</code> where the data cells that the headers apply to are in the same column, row, column group, or row group, respectively.  

Reviewer: ```suggestion
            <li id="12bTI-1b">For any HTML <code>&lt;table&gt;</code>: a header cell may be marked up with <code>&lt;th scope="[col|row|colgroup|rowgroup]"&gt;</code> where the data cells that the headers apply to are in the same column, row, column group, or row group, respectively.  
```

                <ul>

>                     <li>In HTML4, <code>&lt;td scope&gt;</code> is supported.</li>

Reviewer: ```suggestion
                    <li>In HTML4, <code>&lt;td scope="[row|col]"&gt;</code> is supported.</li>
```

>                     <li>In HTML5, <code>&lt;td scope&gt;</code> is not supported so all header cells must be <code>&lt;th&gt;</code>. </li>

Reviewer: ```suggestion
                    <li>In HTML5, <code>&lt;td scope="[row|col]"&gt;</code> is not supported so all header cells must be <code>&lt;th&gt;</code>. </li>
```

                    <li>The <code>scope</code> only applies to cells that occur after the header cell(s) in the reading order.</li>
                </ul>
            </li>

>             <li id="12bTI-1c">For any HTML <code>&lt;table&gt;</code>: each <code>&lt;td&gt;</code> data cell may be associated to header cell(s) with <code>&lt;td headers&gt;</code>:

Reviewer: ```suggestion
            <li id="12bTI-1c">For any HTML <code>&lt;table&gt;</code>: each <code>&lt;td&gt;</code> data cell may be associated to header cell(s) with <code>&lt;td headers="[id]"&gt;</code>:
```

>                 <ul>

Reviewer: Wouldn't the headers technique also apply to ARIA tables? Should we repeat these bullets for the ARIA approach or modify the parent bullet to apply to both HTML tables and ARIA tables?

Author: I don't think so; https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/table_role

`role=rowheader` or `role=columnheader` are to be used in an aria table, grid, or treegrid.

**Baseline Test ID:** 12.A-DataTable
#### Identify Content
<p id="12aIC">Content/data that is visually presented in a table, arranged in rows and columns, where the content is not in a meaningful sequence when linearized.</p>

<p>Note: Linearization of table content is the presentation of a table’s two-dimensional content in one-dimensional order of the content in the source, beginning with the first cell in the first row and ending with the last cell in the last row, from left to right, top to bottom.</p>

#### Test Instructions
<ol id="12aTI">
    <li>In the following instructions, where an ARIA role is identified, it is the first valid ARIA role attribute value.</li>

>     <li id="12aTI-1">Table: Check the data table has programmatic role of table using one of the following techniques. If more than one technique is used, select one and perform the remaining tests using that selection. [SC 1.3.1]:

Author: ```suggestion
    <li id="12aTI-1">Table: Check the data table has programmatic role of table using one of the following techniques. If more than one technique is used, select one and perform the remaining tests using that selection. [SC 4.1.2]:
```

Reviewer 2: Are we using "Check that" or "Check"?
